### PR TITLE
Add system metrics

### PR DESCRIPTION
**Pull request by malafeev**
It is based on https://github.com/open-telemetry/opentelemetry-java/pull/1323
Spec: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/semantic_conventions/system-metrics.md

### CODE REVIEW

**Comment by malafeev** [comment]
cc @carlosalberto

**Comment by iNikem** [comment]
@jkwatson do you want to review this? Metrics are your cup of tea, aren't they?

**Comment by carlosalberto** [comment]
Looks good to me! I'm just wondering where/how should we initialize it, as I'd like this to be part of the rest of the instrumentation.

**Review by carlosalberto — APPROVED** [review]

**Comment by trask** [comment]
hey @malafeev! sorry for the delay, will review this by end of the week

File: system-metrics/system-metrics.gradle
Comment (carlosalberto) [reply system-metrics/system-metrics.gradle:7]: An alternative to this would be to add this but:

1) Disable it by default AND
2) Make oshi a compile dependency.

This way, if users want to leverage it, they'd need to explicitly enable this instrumentation part, AND provide the Oshi dependency themselves.

Comment (anuraaga) [reply system-metrics/system-metrics.gradle:7]: I think it's a good idea to make oshi a compile dependency. How about we model this change exactly like any instrumentation, `instrumentation/oshi/{library,javaagent}` for example. This code goes in library, giving user a manual way of instrumenting (calling the static method) similar to our other library instrumentations, and auto can call it automatically if oshi is found on the classpath (not sure if we have a pattern for this yet but figure it's possible right @trask?). We can start with just moving this into library and add javaagent later.

For reference, https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/master/instrumentation/grpc-1.5/library/grpc-1.5-library.gradle

Comment (anuraaga) [reply system-metrics/system-metrics.gradle:7]: In which case we should split out the JVM metrics into a separate `instrumentation/jvm/library` too

**Comment by iNikem** [comment]
Why this PR is blocked from merging? We do have one approval from @carlosalberto ... @trask do you still want to review this?

File: system-metrics/system-metrics.gradle
Comment (carlosalberto) [reply system-metrics/system-metrics.gradle:7]: > This code goes in library, giving user a manual way of instrumenting (calling the static method) similar to our other library instrumentations

Sounds good to me.

> In which case we should split out the JVM metrics into a separate instrumentation/jvm/library too

Same.

@malafeev any opinion on this? ;)

Comment (malafeev) [reply system-metrics/system-metrics.gradle:7]: I'm fine with it. any other opinions?

**Comment by jkwatson** [comment]
> @jkwatson do you want to review this? Metrics are your cup of tea, aren't they?

Don't hold this up on my account. :)

File: system-metrics/system-metrics.gradle
Comment (trask) [reply system-metrics/system-metrics.gradle:7]: I like the idea of adding just library instrumentation in the first PR, and then work on how to integrate it into javaagent after that

**Comment by trask** [comment]
> Why this PR is blocked from merging?

@iNikem it's blocked from merging because it hasn't been approved by someone in the CODEOWNERS file

File: system-metrics/system-metrics.gradle
Comment (malafeev) [reply system-metrics/system-metrics.gradle]: I moved code to `instrumentation/oshi/library` because both jvm and system metrics depends on oshi

**Comment by malafeev** [comment]
@trask could you review this PR again?

File: instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java
Comment (trask) [inline instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java:21]: maybe rename to `ProcessMetrics` to make clear these are not just standard Java-provided metrics?

File: instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java
Comment (trask) [inline instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java:80]: can this be part of the other (future) module that only depends on JVM-provided metrics?

**Review by trask — COMMENTED** [review]
thx @malafeev! sorry for the delayed review

File: instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java
Comment (malafeev) [reply instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java:80]: It depends on oshi.
If I will move it to separate module it will have the same dependency on oshi as system metrics.
Activation will be the same: presence of oshi dependency.
Therefore I doubt to split it to two modules.

File: instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java
Comment (malafeev) [reply instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java:21]: renamed

File: instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java
Comment (trask) [reply instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java]: `runtime.java.gc_count` metric has dependency on oshi?

Comment (malafeev) [reply instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java]: oh, you mean only this specific metric, I thought you are talking about all jvm metrics here.

Comment (malafeev) [reply instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java]: I can create separate `instrumentation/jvm/libarary` with only this metric.

Comment (trask) [reply instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java]: 👍 how about removing `runtime.java.gc_count` from this PR and then we'll merge this (sorry again it took so long!), and u can send separate PR with `instrumentation/jvm/library`?

Comment (malafeev) [reply instrumentation/oshi/library/src/main/java/io/opentelemetry/instrumentation/oshi/JavaMetrics.java]: removed

**Review by trask — APPROVED** [review]
thx @malafeev 👍

**Comment by trask** [comment]
@malafeev can you merge master into your branch? we updated to otel api 0.10.0-SNAPSHOT, which breaks a couple of metric APIs, just in case it affects this PR

**Comment by malafeev** [comment]
@trask synced with master

**Comment by trask** [comment]
@malafeev looks like the branch needs a `./gradlew spotlessApply`

**Comment by malafeev** [comment]
@trask master needs a `./gradlew spotlessApply` :)

**Comment by trask** [comment]
oh no! I think I messed up with the merge enforcement switching over from CircleCI to GHA, which probably led to that

**Review by anuraaga — APPROVED** [review]
Thanks!

**Pull request merged**
